### ISSUE TITLE
Kong ingress controller creates Kong service with protocol: http instead of https

### ISSUE THREAD

**Issue by suankan** [Kong/kubernetes-ingress-controller#406]
### Summary

My upstream APIs are running over HTTPS.

After upgrading from KIC: 0.1.3, Kong EE: 0.34-1 to KIC 0.5.0, Kong EE: 0.36-2 I'm observing a problem that my API throws error "400 The plain HTTP request was sent to HTTPS port" after KIC published it in Kong.

If I check Kong Service after publishing, I see it is created with setting "protocol: http". 
Before the upgrade Kong Service was created with "protocol: https".

As a result,

**Kong Ingress controller version**
Upgrade: 0.1.3 > 0.5.0

**Kong or Kong Enterprise version**
Upgrade: 0.34-1 > 0.36-2

### Kubernetes version

**paste `kubectl version` output**

### Environment

- **AWS**:
Self-managed Kubernetes deployed on EC2 instances

- **OS**:
RHEL 7 flavour

- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

### What happened

After upgrade my example API throws error "400 The plain HTTP request was sent to HTTPS port" after KIC published it in Kong.

Checking Kong Service publishing, I see it has "protocol: http". Here is the response from Kong Admin after KIC discovered and published my API: 
(I kubcetl port-forwarded my `service/kong-admin`)
```
$ curl -s -k https://localhost:8444/services | jq
{
  "next": null,
  "data": [
    {
      "host": "api-example-001.apig-dev.svc",
      "created_at": 1568949098,
      "connect_timeout": 60000,
      "id": "1b1682bc-ac7f-4cbc-b230-7d16b95a7509",
      "protocol": "http",
      "name": "apig-dev.api-example-001.443",
      "read_timeout": 60000,
      "retries": 5,
      "port": 80,
      "path": "/",
      "updated_at": 1568949098,
      "write_timeout": 60000,
      "tags": [
        "managed-by-ingress-controller"
      ]
    }
  ]
}
```

Before the upgrade Kong Service was created with "protocol: https".

Here are my Ingress and KongIngress resources:
```
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: api-example
  labels:
    app: api-example
  annotations:
    kubernetes.io/ingress.class: kong
    configuration.konghq.com: api-example-001
    kubernetes.io/ingress.allow-http: "false"
spec:
  rules:
    - http:
        paths:
        - backend:
            serviceName: api-example
            servicePort: 443
          path: /api/example/001
---
apiVersion: configuration.konghq.com/v1
kind: KongIngress
metadata:
  name: api-example
  labels:
    app: api-example
proxy:
  protocol: https
route:
  strip_path: false
```

Here is my KongIngress CRD:
```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: kongingresses.configuration.konghq.com
spec:
  group: configuration.konghq.com
  version: v1
  scope: Namespaced
  names:
    kind: KongIngress
    plural: kongingresses
    shortNames:
      - ki
  validation:
    openAPIV3Schema:
      properties:
        route:
          properties:
            methods:
              type: array
              items:
                type: string
            regex_priority:
              type: integer
            strip_path:
              type: boolean
            preserve_host:
              type: boolean
            protocols:
              type: array
              items:
                type: string
                enum:
                  - http
                  - https
        proxy:
          type: object
          properties:
            protocol:
              type: string
              enum:
                - http
                - https
            path:
              type: string
              pattern: ^/.*$
            retries:
              type: integer
              minimum: 0
            connect_timeout:
              type: integer
              minimum: 0
            read_timeout:
              type: integer
              minimum: 0
            write_timeout:
              type: integer
              minimum: 0
        upstream:
          type: object
          properties:
            hash_on:
              type: string
            hash_on_cookie:
              type: string
            hash_on_cookie_path:
              type: string
            hash_on_header:
              type: string
            hash_fallback_header:
              type: string
            hash_fallback:
              type: string
            slots:
              type: integer
              minimum: 10
            healthchecks:
              type: object
              properties:
                active:
                  type: object
                  properties:
                    concurrency:
                      type: integer
                      minimum: 1
                    timeout:
                      type: integer
                      minimum: 0
                    http_path:
                      type: string
                      pattern: ^/.*$
                    healthy: &healthy
                      type: object
                      properties:
                        http_statuses:
                          type: array
                          items:
                            type: integer
                        interval:
                          type: integer
                          minimum: 0
                        successes:
                          type: integer
                          minimum: 0
                    unhealthy: &unhealthy
                      type: object
                      properties:
                        http_failures:
                          type: integer
                          minimum: 0
                        http_statuses:
                          type: array
                          items:
                            type: integer
                        interval:
                          type: integer
                          minimum: 0
                        tcp_failures:
                          type: integer
                          minimum: 0
                        timeout:
                          type: integer
                          minimum: 0
                passive:
                  type: object
                  properties:
                    healthy: *healthy
                    unhealthy: *unhealthy
```

Here is my Service and Deployment for my example api:
```
apiVersion: v1
data:
  nginx-ssl-offloader.conf: |
    server {
        listen                  443 ssl;
        ssl_certificate         /nginx-certs/selfsigned.crt;
        ssl_certificate_key     /nginx-certs/selfsigned.key;
        ssl_protocols           TLSv1 TLSv1.1 TLSv1.2;
        ssl_ciphers             HIGH:!aNULL:!MD5;

        location / {
            proxy_buffering     on;
            proxy_cache         off;

            proxy_set_header    Host $host;
            proxy_set_header    X-Real-IP $remote_addr;
            proxy_set_header    X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header    X-Forwarded-Proto $scheme;

            proxy_pass          http://localhost:8080;
        }

        access_log              /dev/stdout;
        error_log               /dev/stderr error;
    }
kind: ConfigMap
metadata:
  annotations: {}
  labels:
    app: api-example-001
  name: nginx-ssl-offloader-001
---
apiVersion: v1
data:
  selfsigned.crt: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURqekNDQW5lZ0F3SUJBZ0lKQUtwSVFYbUtoU0JsTUEwR0NTcUdTSWIzRFFFQkN3VUFNRjR4Q3pBSkJnTlYKQkFZVEFrRlZNUk13RVFZRFZRUUlEQXBUYjIxbExWTjBZWFJsTVNFd0h3WURWUVFLREJoSmJuUmxjbTVsZENCWAphV1JuYVhSeklGQjBlU0JNZEdReEZ6QVZCZ05WQkFNTURtVmphRzl6WlhKMlpYSXVZMjl0TUI0WERURTVNRGt4Ck9UQTJORGd3TUZvWERUSXdNRGt4T0RBMk5EZ3dNRm93WGpFTE1Ba0dBMVVFQmhNQ1FWVXhFekFSQmdOVkJBZ00KQ2xOdmJXVXRVM1JoZEdVeElUQWZCZ05WQkFvTUdFbHVkR1Z5Ym1WMElGZHBaR2RwZEhNZ1VIUjVJRXgwWkRFWApNQlVHQTFVRUF3d09aV05vYjNObGNuWmxjaTVqYjIwd2dnRWlNQTBHQ1NxR1NJYjNEUUVCQVFVQUE0SUJEd0F3CmdnRUtBb0lCQVFEZEJSV0hkOENDaXA0NDZLRnJkc3BOK3ViWTJRZ0Uyc0pkc0NIWGVzQ1duQ0xhOGJiMTB3dkwKN0M0U1hNdm8yKzRMNE1sd3V2V3RBNVNlQndNdTk0cUNuRForZVVlOVJGNE40aytSQ3JZTVE0eHR6eithSHZjcAo5SC9tQnVhb0VmTjllaUhPTTlicmoySlEvSXMxZXY4bnFnd2ZTOVBWSHVYcUU1SnZuNW5WWUlTNlNTSVlNNmQvCk9nTElCUVJ1c3lwK01IaEdLTUxGVStSUk5kUGl1NzVQUEZhNkk4TnBVVjlaOEhQZDVXMStIbEg1YWh2VG1UKzAKZGMvTHptSGFBYklVdE8yYVIyTjE5em1MNlI0MllUdVUzTTZ4Y0RQUUtnQ0pMbVJsYW5vWUhKRHYrcllZVWs0ZgpaSlNKSDdtZUd6YmtQdXBOWEcxN1FLUDFZeVZubzFpTEFnTUJBQUdqVURCT01CMEdBMVVkRGdRV0JCUjBuQXBsClhSZFlnWXd0YkFjcTBINjczRGc0bkRBZkJnTlZIU01FR0RBV2dCUjBuQXBsWFJkWWdZd3RiQWNxMEg2NzNEZzQKbkRBTUJnTlZIUk1FQlRBREFRSC9NQTBHQ1NxR1NJYjNEUUVCQ3dVQUE0SUJBUUMvUTdlMkY1dzZ2VkFkdGE2NgpQY0FyTlVrUjJjRFlFZmZUOXFNQnRWNWJYTXEwSEJNMlNyTDV0TWdyVEVQRlM4YTNZMWFBN3BRMTBRUDBlWHh2CjRON0pVcDF3RUV6bkd2ODdIVmpXZXRWY081bEVJeWtBVFRDcnpTOTR3TVNrc1NlSlFubFV6OExnUkljTXNDQjIKZzFrZGdYdjF3bjB1WnhaM01KZm9XdU9nVEtNRzdmRzVGb24xc25vd2Z1Vk9QVlZJM0c0ZWszdEd2OGlRcWs2Swo3NkprcUg0c1VMQXJublAvc2UvWFFQSzBtVk9aM1ZRbllJdWY0S0NJZVZVT29FOXFNMGlQaHVUOUtYSGRYelFwCmdETGNsV25SenFPN0wyRzkwQWc3NG9pUmIrOStJYVlyRWtkMkpOanpHZWlMNzk5TTRCMlNBZTNleTREck13VmYKaGFWLwotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
  selfsigned.key: LS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tCk1JSUV2UUlCQURBTkJna3Foa2lHOXcwQkFRRUZBQVNDQktjd2dnU2pBZ0VBQW9JQkFRRGRCUldIZDhDQ2lwNDQKNktGcmRzcE4rdWJZMlFnRTJzSmRzQ0hYZXNDV25DTGE4YmIxMHd2TDdDNFNYTXZvMis0TDRNbHd1dld0QTVTZQpCd011OTRxQ25EWitlVWU5UkY0TjRrK1JDcllNUTR4dHp6K2FIdmNwOUgvbUJ1YW9FZk45ZWlIT005YnJqMkpRCi9JczFldjhucWd3ZlM5UFZIdVhxRTVKdm41blZZSVM2U1NJWU02ZC9PZ0xJQlFSdXN5cCtNSGhHS01MRlUrUlIKTmRQaXU3NVBQRmE2SThOcFVWOVo4SFBkNVcxK0hsSDVhaHZUbVQrMGRjL0x6bUhhQWJJVXRPMmFSMk4xOXptTAo2UjQyWVR1VTNNNnhjRFBRS2dDSkxtUmxhbm9ZSEpEdityWVlVazRmWkpTSkg3bWVHemJrUHVwTlhHMTdRS1AxCll5Vm5vMWlMQWdNQkFBRUNnZ0VBQTBnSkNzclhXOWpOQXBGSm5EdldReFFmaTBhbG5OVFduK2QyaVNsMFpJZ0sKNkc5QVJTUU4vOGdRNHdybnFCeGp5MTlaL2tkNFVFcEdVWlpQRU1weUR3QmN1c0g0TTdJcEoxR2Qxc3h1NlgyegpQd29kcXA4ZEVrMWovTGxMZ1dGSzJTa1FDV3l4bXl1MHpQa0M1QmRnWkdsYzVyWU1maFNJVzFKUkVVRXdXbitlClNKVHBzWjZPZU1qMXBGNFRzSUhXMSt0SFJRVVlVSVhHUTV1UFF1cFQycnVseTA3Q0tOSVBWdFNwRVNQVkNZNnMKM25DRFY5MlJxYTBZSDJ5RTVFclFmaGd2ZERHNW4vSDRtM3dDUmN4ay9WNUJqdE9EQVJvR09FVHRlRlhJTDRvVApBUjRpM1NXdmdGS2liaFh6aVc1YTh4amtIRU11Z1JkUWJrZTVNZS9Qc1FLQmdRRHdhSVBMdFM5SGtzV0VOekZKClM5eksvVEREaHJlci95Qyt1WWc2bkwrd0x5bHV5dlVzTDV5V2k1bHIzbnNYTzVTSWZoWWg3dmdJa0tMTlJveHIKOFJpTjkrQ2ZyVU5KZUtHaks2ZG02VmFiZW9zSWlYU3VlczFleHVTbERZQWltcDRReHloU3JNNUdtV2tmckhyQQovTzRuOXdQN3cxa1U3bGYrMHQxUnF5clpZd0tCZ1FEcldxb3FUdlNRZ2d3eFkra0hCUGV4NjBLL2RPTXdJZi9sCjBpSXdLNFR1QXJ5N2UwVmJuSk5VQkVhKzZFbnR0SW1zeGZMeXNJN3NGMEpiRE5td1BLZXZVWWpwbFZyNU1EZzgKelFVQXdmSjV3SWxPOHRFMHFZd3JTUFZjRll3RGViNEhXQ1MrcEhiM2VNcVZqM1JsZERheForRGxTa1BlSXFtZwpQaHVyMVhiQXVRS0JnRm1FWVVieUFWZmlVLy9HV1dqVWdiTm9FY2RNbTRuRTV4bDlsamRVOUorVEFCZU1KVHk5CndjM2d6MUQyTjlsUHl5QUQ2eFBabXF2OXgxRXVhczBxcWh6NjVyNHVVdmJCbjFaVncvQ3pCQmZDeXBCME5tbGQKdXFrRzRFNkJOMndyTEo1TXVqd0dzb3hoUlY2UTUraUNoZCtGd0FyNEUyeVN6MXRJZ3phTXlOakhBb0dBR2Z0cApXOEJacVhaaDZEWHliTTcwbFZTMmlySlp2bCtTZzRtZU1wa3JWS2RDdEtQWU5WUGt3OXF4QUxSZUxQcklWQ1dLCk81YVlRS0lVdGo2RjdiblVmVHRRRlJGaStLMkZGOWsrV3lZT0liaktyOGVJcHZxc3ViNlZMN1JGUDhENTVJK3IKaHN4YlFkaTQ3S2ZvbnFha3Z4TUx0ZVJyOTZMMS9QVytVa3Iva0NrQ2dZRUFnQXYrRklkOWRpZ3BVenE2UGpMUwpacTBaOXJGWWRIREw5SVBtMDhwYUU1ZytLRGVTSXdNc0ZFdVBXeUpncmRoWUVNTWZKaTA3TmJKN3E3VEE2N05FClBOVk1TTm5hcE55aFlYQnY1TWRmUkZuajIvSGJmZDREUUR2WC9kV05kbWJ5YSt4Z0I4OEdPL0N3bXRZODQrU2sKYVRObFBjUGdwdFVzUi9vN3VqWGMrcTQ9Ci0tLS0tRU5EIFBSSVZBVEUgS0VZLS0tLS0K
kind: Secret
metadata:
  annotations: {}
  labels:
    app: api-example-001
  name: nginx-certs-001
type: Opaque
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: api-example-001
  name: api-example-001
spec:
  ports:
  - name: https
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: api-example-001
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: api-example-001
  name: api-example-001
spec:
  replicas: 1
  selector:
    matchLabels:
      app: api-example-001
  template:
    metadata:
      labels:
        app: api-example-001
    spec:
      containers:
      - env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
        image: docker.artifactory.ai.cba/apig/echoserver:1.8
        name: echoserver
        ports:
        - containerPort: 8080
        resources:
          limits:
            cpu: 100m
            memory: 128Mi
          requests:
            cpu: 100m
            memory: 128Mi
      - image: dockerhub.artifactory.ai.cba/nginx:alpine
        name: nginx
        ports:
        - containerPort: 443
          name: https
        resources:
          limits:
            cpu: 100m
            memory: 128Mi
          requests:
            cpu: 100m
            memory: 128Mi
        volumeMounts:
        - mountPath: /etc/nginx/conf.d
          name: nginx-ssl-offloader
        - mountPath: /nginx-certs
          name: nginx-certs
      volumes:
      - configMap:
          name: nginx-ssl-offloader-001
        name: nginx-ssl-offloader
      - name: nginx-certs
        secret:
          secretName: nginx-certs-001
---
apiVersion: configuration.konghq.com/v1
kind: KongIngress
metadata:
  labels:
    app: api-example-001
  name: api-example-001
proxy:
  protocol: https
route:
  protocols:
  - https
  strip_path: false
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    configuration.konghq.com: api-example-001
    kubernetes.io/ingress.allow-http: "false"
    kubernetes.io/ingress.class: kong
  labels:
    app: api-example-001
  name: api-example-001
spec:
  rules:
  - http:
      paths:
      - backend:
          serviceName: api-example-001
          servicePort: 443
        path: /api/example/001
```

As a result, when I call my published API I get this:
(kubcetl port-forwarded my `service/kong-proxy`)
```
$ curl -k https://localhost:8443/api/example/001
<html>
<head><title>400 The plain HTTP request was sent to HTTPS port</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<center>The plain HTTP request was sent to HTTPS port</center>
<hr><center>nginx/1.17.3</center>
</body>
</html>
```

At the same time if I call the same API directly via my K8s Service I can get a successful response:
(kubcetl port-forwarded my `service/api-example-001`)
```
$ curl -k https://localhost:8443/api/example/001


Hostname: api-example-001-5cc9794d54-llxnr

Pod Information:
	node name:	ip-10-8-214-124.ap-southeast-2.compute.internal
	pod name:	api-example-001-5cc9794d54-llxnr
	pod namespace:	apig-dev
	pod IP:	10.8.214.81

Server values:
	server_version=nginx: 1.13.3 - lua: 10008

Request Information:
	client_address=127.0.0.1
	method=GET
	real path=/api/example/001
	query=
	request_version=1
	request_uri=http://localhost:8080/api/example/001

Request Headers:
	accept=*/*
	connection=close
	host=localhost
	user-agent=curl/7.54.0
	x-forwarded-for=127.0.0.1
	x-forwarded-proto=https
	x-real-ip=127.0.0.1

Request Body:
	-no body in request-

```

### Expected behaviour

HTTPS API should be available after KIC published it in Kong.

### Steps To Reproduce

Please see above.

**Comment by suankan** [comment]
# Update

I've found this forum thread that resolved my problem.
https://discuss.konghq.com/t/ingress-controller-registers-https-services-as-http-and-port-443-as-port-80/2784/7

TLDR;

The suggestion was to add annotation **configuration.konghq.com: api-example-001** to the Service referring to corresponding KongIngress resource:
```
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    configuration.konghq.com: api-example-001
  labels:
    app: api-example-001
  name: api-example-001
spec:
  ports:
  - name: https
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: api-example-001
```
